src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportDestinationClient/ExportDestinationClientFactory.cs

```diff
@@ -0,0 +1,46 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using EnsureThat;
+
+namespace Microsoft.Health.Fhir.Core.Features.Operations.Export.ExportDestinationClient
+{
+    /// <summary>
+    /// Provides functionality to create an instance of <see cref="IExportDestinationClient"/> based on destination type.
+    /// </summary>
+    public class ExportDestinationClientFactory : IExportDestinationClientFactory
+    {
+        private Dictionary<string, Func<IExportDestinationClient>> _registeredTypes;
+
+        public ExportDestinationClientFactory(IEnumerable<Func<IExportDestinationClient>> destinationClientFactories)
+        {
+            EnsureArg.IsNotNull(destinationClientFactories, nameof(destinationClientFactories));
+
+            _registeredTypes = destinationClientFactories.ToDictionary(factory => factory().DestinationType, StringComparer.Ordinal);
```

Comment:
I don't like the fact this has to instantiate an instance of the factory to get the `DestinationType` so that it can register the mapping. I think a better way would be to use Attribute or something against the class so that it would be treated more like a metadata of the class. However, that would require some work in the IoC registration process to support that. For now, this works and I will refactor the code later.